core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/federation/RepositoryFederatedService.java

```diff
@@ -76,8 +85,27 @@ public BatchingServiceIteration(CloseableIteration<BindingSet> inputBindings,
 
 		@Override
 		protected void handleBindings() throws Exception {
-			while (!isClosed() && leftIter.hasNext()) {
+			// Set up a consumer task to send HTTP requests in parallel. This must be done in a
+			// separate thread, because submitting HTTP requests may block if the HTTP pool is full.
+			// In that case, we would enter a deadlock, with the main thread waiting for both the
+			// pool to yield, and the consumer of the bindings to read from the queue.
+			// See: https://github.com/eclipse-rdf4j/rdf4j/discussions/5120
+			// Test case: https://github.com/tkuhn/rdf4j-timeout-test
+			final Runnable queueConsumer = () -> {
+				requestQueue.forEachRemaining(batchIter -> {
+					// evaluateInternal is BLOCKING if the HTTP pool is exhausted
+					addResult(evaluateInternal(service, batchIter, service.getBaseURI()));
+				});
+			};
+			if (queueConsumerTask != null) {
+				// This should never happen, but just to be sure...
+				throw new QueryEvaluationException("Queue consumer task already running");
+			}
+			// This may throw, we simply let it propagate
+			queueConsumerTask = ForkJoinPool.commonPool().submit(queueConsumer);
 
+			// Produce batches of input bindings and send them to the consumer via the request queue
+			while (!isClosed() && leftIter.hasNext()) {
```

Comment:
Would it make sense to check if the current thread is interrupted? To make it easier to cancel a query that is stuck. Unless this is already handled elsewhere in such a way that we don't need to check here.

Reply:
Interruption state is checked in `requestQueue.put(...)` and then also handled in `JoinExecutorBase`. This should be fine, I think.